runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -57,6 +60,13 @@
   private final ParDoEvaluatorFactory<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
       delegateFactory;
   private final EvaluationContext evaluationContext;
+  private final Collection<DoFnLifecycleManager> doFnLifecycleManagers =
```

Comment:
I'm worried about caching the `DoFnLifecycleManager` from every bundle of a `SplittableDoFn` over the entire lifecycle of the pipeline - that'll grow without bound for any unbounded SDF, which we definitely don't want. We either need to reuse these per `AppliedPTransform` (as we do in the base `ParDoEvaluator`), or clean them up whenever we complete a bundle.

Reply:
Id be to clean them up with bundles but also to keep that for when SDF will really be worked on and usable by most runners. Today it is not the case yet so not that critical for this PR (dont want that small fix become an "impl SDF lifecycle" which is way more work and discussion probably)

Reply:
If you're not going to clean up these managers on bundle completion you should just copy the code for caching from the `ParDoEvaluator` that handles reusing DoFns so we can cache the instantiated process fns. We shouldn't let this grow without bound.

Reply:
This is what doesn't work until we recycle them due to the commitbundle dependency, no? Happy to get help on this one if you can @tgroh

Reply:
I'm not sure what you mean when you say 'commitbundle dependency'. Can you elaborate?

Reply:
If you check the unstantiation, it depends on the commitbundle which changes so we cant cache it, can we?